x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/ValuesExtractor.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.mapreduce;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSource.Bytes;
+import org.elasticsearch.search.aggregations.support.ValuesSource.Numeric;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Interface to extract values from Lucene in order to feed it into the MapReducer.
+ */
+public abstract class ValuesExtractor {
```

Comment:
This is heavily inspired by the multi terms aggregation (see `MultiTermsAggregator`). The main difference: In `frequent_items` we care about the value _and_ the field, multi terms throws everything into one bag.

If you have concrete suggestions how this can be simplified I am happy to hear them. I think what might confuse you is naming. It uses the value source registry, see `MapReduceValueSourceRegistry`. The difference to multi terms: it isn't bunched together in one file.

So the entry point you should look at is `MapReduceValueSourceRegistry`, the `ValuesExtractor` - what is called `TermValues` in multi terms - wraps the necessary translation from lucene. It could also be implemented as big `if`. As said, the name  `ValuesExtractor` isn't that great, especially as it collects fields(keys) and values, not just values.

TL/DR Using field id's, instead of field names and resolving the id's to names at final reduce is a big memory saver.

Reply:
Echoing my earlier comment, MultiTerms makes it clear that those KeyConverters are specific to MultiTerms.  This looks like it's trying to be generic.  Again, I'd be more okay with this if it were tightly bound to the specific aggregation it's supporting.